docs/upgrade.html

```diff
@@ -61,10 +62,10 @@ <h4><a id="upgrade_2_0_0" href="#upgrade_2_0_0">Upgrading from 0.8.x, 0.9.x, 0.1
         Similarly for the message format version.</li>
     <li>If you are using Java8 method references in your Kafka Streams code you might need to update your code to resolve method ambiguities.
         Hot-swapping the jar-file only might not work.</li>
-    <li>ACLs should not be added to prefixed resources,
+    <li>ACLs should not be added during the upgrade process.
```

Comment:
Also, if we have upgraded the whole cluster and started using the new acl_change format in ZK, downgrading the cluster will cause the processing of acl_change sequence node to fail and the sequence node not to be deleted. This means the downgraded broker can't process ACL changes forever. So, your original idea of using a new acl_change path such as /kafka-acl-changes-v2 seems better.

Reply:
Actually, this has the other problem. Since the AclCommand tool still writes to ZK directly, if we switch to a new acl-change path, the old version of AclCommand won't work once the brokers are upgraded.

Given the complexity in dealing with upgrades/downgrades, maybe the current approach of using 2 separate acl-change path is not that bad. So, perhaps we can just keep the current approach, but (1) ensure that the new acl with prefix is only added if inter.broker.protocol is new; (2) In ZkNodeChangeNotificationListener.processNotifications(), catch decoding error of individual acl change value, log it and let it go. This may help future format change.

Sorry for going back and forth on this one.

Reply:
Hummm.... Good points.

I have another suggestion...

pre-2.0 behaviour uses a colon-separated string to write the change notification in the format '<resource-type>:<name>', where the name can have embedded colons, e.g. a group of 'my:weird:group:name'. 

With my recent changes the toString representation of `Resource` is '<name-type>:<resource-type>:<name>', and I've already enhanced the `Resource.fromString()` method to handle both old and new formats. For the solution below I'll need to swap the first two parts in the new format so that it is '<resource-type>:<name-type>:<name>'.

Given this new format, rather than  switching to JSON we could just write out change notifications out as strings still, with either two or three parts, depending on inter.broker.protocol.version.  

During a rolling upgrade change events would be written using two parts, which work with older brokers and are interpreted as 'literal' resources by the new brokers. (Attempts to add Prefixed would result in UVE).

Once upgraded brokers start to write using three parts change events.

On a downgrade older brokers would treat the 'Topic:Prefixed:Foo' as a change on a Topic called 'Prefixed:Foo', which wouldn't match anything.  I'd have to check, but there's every possibility that the broker would process this without error and move on.  The fact that the old broker effectively ignored the event shouldn't matter as the event must be old to be in this format.

Thoughts?

Reply:
Yes, that seems to be a good idea. The only confusion is that if someone has a group named "Literal:weirdgroup" or "Prefixed:weirdgroup". Since that's unlikely, I think what you suggested is probably the best approach given what we have now.